memtierd: redesign command line arguments

askervin opened the pull request:
- Introduce three ways to run memtierd commands, any combination of which is allowed and makes perfect sense. In the order of execution the ways are: command line arguments (-c COMMANDS), command file (-f FILE) and the interactive prompt (--prompt).
- Show interactive prompt only if requested by the user (--prompt).
- Add logging into a file (-l FILE).
- Add option for echoing executed commands from any source (--echo).
- Drop --config-dump.
- Update the command line parameters with which memtierd is launched in e2e tests.

askervin commented:
Depends on PR #66

yanjing1104 commented on cmd/memtierd/main.go:
```diff
@@ -75,39 +75,62 @@ func loadConfigFile(filename string) (memtier.Policy, []memtier.Routine) {
+	// Otherwise (no interactive prompt or it is not quitting) if policy
+	// or routines are configured, memtierd will not exit.
+
+	if *optConfig == "" && *optCommandString == "" && *optCommandFile == "" && *optPrompt == false {
```
maybe simply "!*optPrompt"?

askervin replied:
Thanks, fixed.

yanjing1104 commented on cmd/memtierd/main.go:
```diff
@@ -75,39 +75,62 @@ func loadConfigFile(filename string) (memtier.Policy, []memtier.Routine) {
-	memtier.SetLogger(log.New(os.Stderr, "", 0))
-	optPrompt := flag.Bool("prompt", false, "launch interactive prompt (ignore other parameters)")
-	optConfig := flag.String("config", "", "launch non-interactive mode with config file")
-	optConfigDumpJSON := flag.Bool("config-dump-json", false, "dump effective configuration in JSON")
```
since --config-dump is dropped, the related e2e tests should be modified accordingly.

- memtierd.test-suite/no-policy/n4c16/test01-mover/code.var.sh:    '-config-dump'
- memtierd.test-suite/no-policy/n4c16/test02-tracker-finder/code.var.sh:memtierd-command "tracker -create finder -config {\\\"ReportAccesses\\\":42} -config-dump -start ${MEME_PID} -counters"
- memtierd.test-suite/no-policy/n4c16/test02-tracker-finder/code.var.sh:memtierd-command "tracker -create finder -config {\\\"PagemapBitPresent\\\":false} -config-dump -start ${MEME_PID} -counters"
- memtierd.test-suite/no-policy/n4c16/test02-tracker-finder/code.var.sh:memtierd-command "tracker -create finder -config {\\\"PagemapBitPresent\\\":true} -config-dump -start ${MEME_PID} -counters"
- memtierd.test-suite/no-policy/n4c16/test02-tracker-finder/code.var.sh:memtierd-command "tracker -create finder -config \"{\\\"ReportAccesses\\\":2, \\\"PagemapBitSoftDirty\\\":true}\" -config-dump -start ${MEME_PID} -counters"

askervin replied:
I think we did not have test for `-config-dump` for the memtierd command line (the main function). The main program `-config-dump` dumped the configuration file that was given as `memtierd -config file.yaml`.

The `-config-dump` parameters in the tests are handled by `cmd*` functions in prompt.go. Those dumps are not affected by this PR.

askervin commented:
Thanks for the review, Jing!

I rebased this PR, as the dependency #66 was merged. Now this should build.

I fixed one comment and marked it resolved. The other issue I only explained to you. Please mark the discussion Resolved, if the explanation is good enough for you and Approve the PR. Otherwise, let's continue discussion.
